File: functions/src/cbam/report/v6/jcs.ts
/**
 * G-15 — RFC 8785 JSON Canonicalization Scheme (JCS).
 *
 * Canonical serialisation of hash input. Every hash entering the sealed
 * package is produced through this module so the same logical content yields
 * the identical byte sequence in every runtime (Node.js, Python, Go).
 *
 * RFC 8785 rules applied here:
 *  - Object keys ordered by Unicode code point (not UTF-16 code unit).
 *  - Strings serialised with JSON escaping plus \u2028/\u2029 escapes.
 *  - Numbers serialised per ECMAScript Number (shortest round-trip), which
 *    equals JSON.stringify for finite values.
 *  - No whitespace between tokens.
 *
 * Package-specific exception (G-16): monetary, emissions and ratio values are
 * carried as string-encoded fixed-point decimals and therefore never reach
 * this module as IEEE 754 numbers. undefined is serialised as the literal
 * null, matching the historical package rule (HASH_CANONICAL_RULE 1.0).
 */
import crypto from "node:crypto";

export type JcsValue = null | boolean | number | string | JcsValue[] | { [key: string]: JcsValue };

export const JCS_RULE_ID = "RFC8785:JCS:1.0";

function escapeJsonString(value: string): string {
  let out = JSON.stringify(value);
  if (out.includes("\u2028")) out = out.replaceAll("\u2028", "\\u2028");
  if (out.includes("\u2029")) out = out.replaceAll("\u2029", "\\u2029");
  return out;
}

function compareUnicodeCodePoints(left: string, right: string): number {
  const max = Math.min(left.length, right.length);
  for (let index = 0; index < max; index += 1) {
    const leftPoint = left.codePointAt(index);
    const rightPoint = right.codePointAt(index);
    if (leftPoint === undefined || rightPoint === undefined) break;
    if (leftPoint !== rightPoint) return leftPoint < rightPoint ? -1 : 1;
    if (leftPoint > 0xffff) index += 1;
  }
  return left.length - right.length;
}

export function canonicalJcs(value: unknown): string {
  if (value === undefined || value === null) return "null";
  if (typeof value === "string") return escapeJsonString(value);
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") {
    if (!Number.isFinite(value)) return "null";
    return String(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => canonicalJcs(item)).join(",")}]`;
  }
  const record = value as Record<string, unknown>;
  const keys = Object.keys(record).sort(compareUnicodeCodePoints);
  return `{${keys.map((key) => `${escapeJsonString(key)}:${canonicalJcs(record[key])}`).join(",")}}`;
}

export function reproduceJcsHash(value: unknown): string {
  return crypto
    .createHash("sha256")
    .update(canonicalJcs(value), "utf8")
    .digest("hex");
}

/**
 * G-16 — runtime guard. Walks a hash-bound structure and refuses any IEEE 754
 * number in a field whose name carries a monetary/emissions/rate meaning.
 * Non-monetary metadata fields (indices, counts, sizes) are allowed to remain
 * numbers; the ESLint rule (scripts/lint/no-float-in-hashed-fields.mjs) blocks
 * number literals at the call site.
 */
const FLOAT_BEARING_FIELD =
  /(?:emission|embedded|direct|indirect|grid|electricity|amount|price|cost|value|total|rate|factor|share|ratio|conversion|quantity|volume)/i;

export function assertNoFloatFieldsInHash(value: unknown, path = "$"): void {
  if (value === null || value === undefined) return;
  if (typeof value === "number" && Number.isFinite(value)) {
    throw new Error(`FLOAT_IN_HASH_FIELD:${path}`);
  }
  if (typeof value === "boolean" || typeof value === "string") return;
  if (Array.isArray(value)) {
    for (let index = 0; index < value.length; index += 1) {
      assertNoFloatFieldsInHash(value[index], `${path}[${index}]`);
    }
    return;
  }
  if (typeof value === "object") {
    for (const [key, child] of Object.entries(value as Record<string, unknown>)) {
      if (typeof child === "number") {
        if (FLOAT_BEARING_FIELD.test(key)) {
          throw new Error(`FLOAT_IN_HASH_FIELD:${path}.${key}`);
        }
        continue;
      }
      assertNoFloatFieldsInHash(child, `${path}.${key}`);
    }
  }
}

File: scripts/verify/jcs.py
#!/usr/bin/env python3
"""G-15 / G-20 — RFC 8785 JSON Canonicalization Scheme (JCS), Python runtime.

Byte-identical to functions/src/cbam/report/v6/jcs.ts: object keys are ordered
by Unicode code point, strings use JSON escaping plus \\u2028/\\u2029 escapes,
numbers use the ECMAScript shortest round-trip representation, and no
whitespace is emitted. A structure hashed here yields the same SHA-256 as the
Node implementation.

CLI modes (used by the cross-runtime gate tests):
  echo '<json>' | python3 jcs.py --canonical   # print canonical serialisation
  echo '<json>' | python3 jcs.py --hash        # print sha256(canonical utf-8)
"""
import hashlib
import json
import sys


def _escape_json_string(value: str) -> str:
    out = []
    for ch in value:
        cp = ord(ch)
        if ch == '"':
            out.append('\\"')
        elif ch == "\\":
            out.append("\\\\")
        elif ch == "\b":
            out.append("\\b")
        elif ch == "\f":
            out.append("\\f")
        elif ch == "\n":
            out.append("\\n")
        elif ch == "\r":
            out.append("\\r")
        elif ch == "\t":
            out.append("\\t")
        elif cp < 0x20 or cp == 0x2028 or cp == 0x2029:
            out.append("\\u%04x" % cp)
        else:
            out.append(ch)
    return '"' + "".join(out) + '"'


def _js_number_to_string(value: float) -> str:
    """ECMAScript Number->String (shortest round-trip, JSON.stringify form).

    Integers below 1e21 print in decimal; finite non-integers and values at or
    beyond 1e21 use the shortest round-trip repr (Python's repr matches the
    ECMAScript result for every value the package hashes under G-16, where
    monetary/emissions values are always string-encoded fixed-point decimals).
    """
    if value != value or value in (float("inf"), float("-inf")):
        return "null"
    if value == 0:
        return "0"
    if value.is_integer() and abs(value) < 1e21:
        return str(int(value))
    return repr(value)


def canonical_jcs(value: object) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, str):
        return _escape_json_string(value)
    if isinstance(value, float):
        return _js_number_to_string(value)
    if isinstance(value, int):
        return str(value)
    if isinstance(value, (list, tuple)):
        return "[" + ",".join(canonical_jcs(item) for item in value) + "]"
    if isinstance(value, dict):
        keys = sorted(value.keys())
        return "{" + ",".join(
            _escape_json_string(key) + ":" + canonical_jcs(value[key]) for key in keys
        ) + "}"
    raise TypeError(f"unsupported JCS value type: {type(value).__name__}")


def reproduce_jcs_hash(value: object) -> str:
    return hashlib.sha256(canonical_jcs(value).encode("utf-8")).hexdigest()


def main() -> None:
    mode = "--hash" if "--hash" in sys.argv else ("--canonical" if "--canonical" in sys.argv else None)
    if mode is None:
        print("usage: python3 jcs.py --canonical|--hash < json", file=sys.stderr)
        sys.exit(2)
    raw = sys.stdin.read()
    try:
        value = json.loads(raw)
    except json.JSONDecodeError as err:
        print(f"INVALID_JSON: {err}", file=sys.stderr)
        sys.exit(2)
    if mode == "--canonical":
        print(canonical_jcs(value))
    else:
        print(reproduce_jcs_hash(value))


if __name__ == "__main__":
    main()
